unic-langid-impl/src/tinystr.rs

```diff
@@ -0,0 +1,369 @@
+//! A small ASCII-only bounded length string representation.
+
+use std::cmp::Ordering;
+use std::fmt;
+use std::num::{NonZeroU32, NonZeroU64};
+use std::ops::Deref;
+use std::ptr::copy_nonoverlapping;
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum Error {
+    InvalidSize,
+    InvalidNull,
+    NonAscii,
+}
+
+/// A tiny string that is from 1 to 8 non-NUL ASCII characters.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct TinyStr8(NonZeroU64);
+
+/// A tiny string that is from 1 to 4 non-NUL ASCII characters.
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct TinyStr4(NonZeroU32);
+
+impl TinyStr8 {
+    pub fn new_unchecked(text: &str) -> TinyStr8 {
+        unsafe {
+            let mut word: u64 = 0;
+            copy_nonoverlapping(text.as_ptr(), &mut word as *mut u64 as *mut u8, text.len());
```

Comment:
I think it might be possible to do this without unsafe, using a `[u8; 8]` buffer and then [u64::from_ne_bytes](https://doc.rust-lang.org/std/primitive.u64.html#method.from_ne_bytes), which is present as of 1.32.

Reply:
Hmmm, I'm trying to do that, but I failed to find a way to get `[u8; 8]` out of a `&str`. What API should I use?

Reply:
nevermind, I'll actually go with dumping numbers and `new_unchecked` will take it back.